### PR TITLE
fix out-of-bounds write in sm2_crypt.c

### PR DESCRIPTION
**Pull request by ymlbright**
asn1_encode has two form length octets: short form(1 byte), long form(1+n byte).

CLA: Trivial

detail see: http://luca.ntop.org/Teaching/Appunti/asn1.html
previous PR #6694 doesn't fix this completely.
In patch, we assume that msg_len is smaller than 3GB.
There is a example that makes out-of-bounds write:

```
30 81 8a 02 21 00 e0 1f
84 8c 55 83 2c dd b1 69
50 09 92 a3 92 c0 49 06
c5 80 97 45 01 02 36 ae
c9 67 b1 41 ec a4 02 21
00 86 1b a3 9e eb 59 41
6a f5 12 39 3c a9 24 2c
08 f6 60 2f 5c 5a f9 88
63 6e e6 e7 22 13 cd 20
67 04 20 11 2d ed 9b dc
ca 82 f2 d8 a0 f9 38 92
6f 1b f3 85 8f 82 7d 78
3e 52 8c 24 d5 4c aa 8a
33 a4 67 04 20 e5 1f 47
99 f8 a1 81 77 c7 ff 80
e5 29 23 b7 58 c7 91 7b
6b aa 31 41 74 2d 30 15
5b 40 37 67 4d  
```

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/sm2/sm2_crypt.c [hunk 1]
```diff
@@ -95,7 +95,7 @@ int sm2_ciphertext_size(const EC_KEY *key, const EVP_MD *digest, size_t msg_len,
     if (field_size == 0 || md_size < 0)
         return 0;
 
-    *ct_size = 12 + 2 * field_size + (size_t)md_size + msg_len;
+    *ct_size = 20 + 2 * field_size + (size_t)md_size + msg_len;
```
Comment (mattcaswell) [inline crypto/sm2/sm2_crypt.c]: So why 20? I'm not a fan of "magic" numbers. It will be useful to have a comment explaining where these numbers come from.

Comment (kroeckx) [reply crypto/sm2/sm2_crypt.c]: The function above also has such magic number 10, and this was also 10 initially.

Comment (ymlbright) [reply crypto/sm2/sm2_crypt.c]: In long form length octets, there is how 20 comes from:
[Sequence Tag, 1 byte] + [Sequence Length Size, 1 byte] + [Sequence Length, max to 4 byte*]
[Integer Tag, 1 byte] + [Integer Length, 1 byte] + [00 prefix, 1 byte] + [X Data]
[Integer Tag, 1 byte] + [Integer Length, 1 byte] + [00 prefix, 1 byte] + [Y Data]
[String Tag, 1 byte] + [String Length, 1 byte] + [C3 Data]
[String Tag, 1 byte] + [String Length Size, 1 byte] + [String Length, max to 4 byte*] + [C2 Data]

> * means, we assume that data in field is smaller than 3GB

So, 6+3+3+2+6 = 20

Comment (paulidale) [reply crypto/sm2/sm2_crypt.c]: I don't see why 20 either.  A 3GB size means four bytes of length so I could understand 16.  Where is the other 4 coming from?

Comment (ymlbright) [reply crypto/sm2/sm2_crypt.c]: "Sequence Length" is the length of all followed data, from [Integer Tag, 1 byte] to [C2 Data]. Sequence Length is four bytes when C2 Data is four bytes.

Comment (paulidale) [reply crypto/sm2/sm2_crypt.c]: What about using `ASN1_object_size` to calculate the true length ?

Comment (paulidale) [reply crypto/sm2/sm2_crypt.c]: Thus:

```
    size_t sz;

    sz = 2 * ASN1_object_size(0, field_size + 1, V_ASN1_OCTET_STRING)
         + ASN1_object_size(0, md_size, V_ASN1_INTEGER)
         + ASN1_object_size(0, msg_len, V_ASN1_INTEGER);
    *ct_size = ASN1_object_size(0, sz, V_ASN1_SEQUENCE);
```

Comment (ymlbright) [reply crypto/sm2/sm2_crypt.c]: Sounds great!  but it should be:

Comment (ymlbright) [reply crypto/sm2/sm2_crypt.c]: Thanks, that's great! but md and msg are string, x and y field are integer. I think it should be:
```
size_t sz;

sz = 2 * ASN1_object_size(0, field_size + 1, V_ASN1_INTEGER)
         + ASN1_object_size(0, md_size, V_ASN1_OCTET_STRING)
         + ASN1_object_size(0, msg_len, V_ASN1_OCTET_STRING);
*ct_size = ASN1_object_size(0, sz, V_ASN1_SEQUENCE);
```

PR is updated

Comment (paulidale) [reply crypto/sm2/sm2_crypt.c]: Apologies for getting the types backwards.

Something similar might be possible for the case above that starts `10 +`

File: crypto/sm2/sm2_crypt.c [hunk 2]
```diff
@@ -91,11 +91,16 @@ int sm2_ciphertext_size(const EC_KEY *key, const EVP_MD *digest, size_t msg_len,
...
+    size_t sz;
 
     if (field_size == 0 || md_size < 0)
         return 0;
 
-    *ct_size = 12 + 2 * field_size + (size_t)md_size + msg_len;
+    sz = 2 * ASN1_object_size(0, field_size + 1, V_ASN1_INTEGER)
+         + ASN1_object_size(0, md_size, V_ASN1_OCTET_STRING)
+         + ASN1_object_size(0, msg_len, V_ASN1_OCTET_STRING);
+    *ct_size = ASN1_object_size(0, sz, V_ASN1_SEQUENCE);
```
[the remark this answers is not in the document]
Comment (paulidale) [reply crypto/sm2/sm2_crypt.c]: Thanks @mattcaswell the lack of documentation was frustrating.
My approval stands with this change incorporated.

**Comment by mattcaswell** [comment]
The Travis red cross is relevant (use `/*...*/` comment style not `//`)

**Comment by romen** [comment]
>The Travis red cross is relevant (use `/*...*/` comment style not `//`)

Thanks @mattcaswell for spotting that!

I reconfirm my approval.

**Comment by paulidale** [comment]
Still approved.

Merged, thanks.

**Pull request closed without merging**